api/client/client_test.go

```diff
@@ -0,0 +1,487 @@
+
+	storeID, err := client.NewStore()
+	checkErr(t, err)
+	err = client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+	existingPerson := createPerson()
+	err = client.ModelCreate(storeID, modelName, existingPerson)
+	checkErr(t, err)
+
+	txn, err := client.WriteTransaction(storeID, modelName)
+	if err != nil {
+		t.Fatalf("failed to create write txn: %v", err)
+	}
+
+	defer func() {
+		err = txn.End()
+		if err != nil {
+			t.Fatalf("failed to end txn: %v", err)
+		}
+	}()
+
+	err = txn.Start()
+	if err != nil {
+		t.Fatalf("failed to start write txn: %v", err)
+	}
+
+	person := createPerson()
+
+	err = txn.Create(person)
+	if err != nil {
+		t.Fatalf("failed to create in write txn: %v", err)
+	}
+	if person.ID == "" {
+		t.Fatalf("expected an entity id to be set but it wasn't")
+	}
+
+	has, err := txn.Has(existingPerson.ID)
+	if err != nil {
+		t.Fatalf("failed to write txn has: %v", err)
+	}
+	if !has {
+		t.Fatalf("expected has to be true but it wasn't")
+	}
+
+	foundExistingPerson := &Person{}
+	err = txn.FindByID(existingPerson.ID, foundExistingPerson)
+	if err != nil {
+		t.Fatalf("failed to txn find by id: %v", err)
+	}
+	if !reflect.DeepEqual(foundExistingPerson, existingPerson) {
+		t.Fatalf("txn model found by id does't equal the original")
+	}
+
+	existingPerson.Age = 99
+	err = txn.Save(existingPerson)
+	if err != nil {
+		t.Fatalf("failed to save in write txn: %v", err)
+	}
+
+	err = txn.Delete(existingPerson.ID)
+	if err != nil {
+		t.Fatalf("failed to delete in write txn: %v", err)
+	}
+}
+
+func TestListen(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, err := client.NewStore()
+	checkErr(t, err)
+	err = client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+
+	person := createPerson()
+
+	err = client.ModelCreate(storeID, modelName, person)
+	checkErr(t, err)
+
+	channel, discard, err := client.Listen(storeID, modelName, person.ID)
+	if err != nil {
+		t.Fatalf("failed to listen: %v", err)
+	}
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		person.Age = 30
+		_ = client.ModelSave(storeID, modelName, person)
+		person.Age = 40
+		_ = client.ModelSave(storeID, modelName, person)
+	}()
+
+	val, ok := <-channel
+	if !ok {
+		t.Fatal("channel no longer active at first event")
+	} else {
+		p := &Person{}
+		if err := json.Unmarshal(val, p); err != nil {
+			t.Fatalf("failed to unmarshal listen result: %v", err)
+		}
+		if p.Age != 30 {
+			t.Fatalf("expected listen result age = 30 but got: %v", p.Age)
+		}
+	}
+
+	val, ok = <-channel
+	if !ok {
+		t.Fatal("channel no longer active at second event")
+	} else {
+		p := &Person{}
+		if err := json.Unmarshal(val, p); err != nil {
+			t.Fatalf("failed to unmarshal listen result: %v", err)
+		}
+		if p.Age != 40 {
+			t.Fatalf("expected listen result age = 40 but got: %v", p.Age)
+		}
+	}
+	discard()
```

Comment:
We call the listener discard method here which should cancel the underlying gRPC request that is consuming the server stream. Even though I think that is working, it seems that the `defer clean()` call is hanging up for some reason, possibly related to the gRPC listen request status or something. It's hard to figure out, could use some help here.